[WIP] Migration guide for providers to implement V2 primitives (#2496)
Qiskit/documentation: docs/migration-guides/qiskit-runtime.mdx

- Because both `backend.run()` and the "version 1" primitives are being deprecated, this guide uses only the V2 primitives.
</Admonition>

> The `qiskit-ibm-runtime` package provides cloud access to the IBM&reg; QPUs (quantum processing units) through the primitives interface. The `backend.run()` interface coexisted with the original (V1) primitives model as the dedicated “direct hardware access” entry point.  With the introduction of the V2 primitives interface, the new `SamplerV2` class now fulfills that role. Consequentially, `backend.run()` is being deprecated, along with `qiskit-ibm-provider`, which only exposed the `backend.run()` interface.

Reviewer: ```suggestion
The `qiskit-ibm-runtime` package provides cloud access to the IBM&reg; QPUs (quantum processing units) through the primitives interface. The `backend.run()` interface coexisted with the original (V1) primitives model as the dedicated “direct hardware access” entry point.  With the introduction of the V2 primitives interface, the new `SamplerV2` class now fulfills that role. Consequentially, `backend.run()` and `qiskit-ibm-provider`, which only exposed the `backend.run()` interface, are no longer supported.
```

## Why migrate to Qiskit Runtime primitives?

Similar to the early days of classical computers, when developers had to manipulate CPU registers directly,
the early interface to QPUs simply returned the raw data coming out of the control electronics.
This was not a huge issue when QPUs lived in labs and only allowed direct access by researchers.
Acknowledging that most developers would not and should not be familiar with distilling such raw data into
0s and 1s, IBM introduced `backend.run`, a first abstraction for accessing QPUs in the cloud. This allowed developers
to operate on a familiar data format and focus on the bigger picture.

As access to QPUs became more widespread, and with more quantum algorithms being developed,
again the need for a higher-level abstraction emerged. In response, IBM introduced
the Qiskit [primitives interface](/guides/primitives), which are optimized for two core tasks in quantum algorithm development:

> expectation value estimation (Estimator) and circuit sampling (Sampler). The goal is once

Reviewer: ```suggestion
expectation value estimation (`Estimator`) and circuit sampling (`Sampler`). The goal is once
```

again to help developers to focus more on innovation and less on data conversion.

> For backward compatibility, the `backend.run` interface continues to exist in Qiskit. However, it is deprecated
> in Qiskit Runtime, as most IBM Quantum&reg; users have migrated to V2 primitives due to their improved

Reviewer: ```suggestion
For backward compatibility, the `backend.run` interface continues to exist in Qiskit. However, it is no longer supported
in Qiskit Runtime after the deprecation period, as most IBM Quantum&reg; users have migrated to V2 primitives due to their improved
```

- [Migrate from `qiskit-ibm-provider`](qiskit-runtime-from-ibm-provider)
- [Migrate from `qiskit-ibmq-provider`](qiskit-runtime-from-ibmq-provider)

> ### Step 3: Replace the call to `backend.run` with a call to primitives.

Reviewer 2: ```suggestion
### Step 3: Replace the call to `backend.run` with a call to primitives
```

See these topics for instructions if you are using `qiskit-ibm-runtime`:

-  [Update code that performs circuit sampling](qiskit-runtime-examples#sampler-algorithm)
-  [Update code that calculates expectation values](qiskit-runtime-examples#estimator-algorithm)
-  [Common use cases (basic, parameterized, and dynamic circuits)](qiskit-runtime-use-case)
-  Replace any `backend.run` options with `qiskit_ibm_runtime` options:
    - [Migrate options](qiskit-runtime-options)
    - Common use cases [Options section](qiskit-runtime-use-case#migrate-options)

> See the following topic for instructions if you are using a 3rd-party provider that only exposes `backend.run`:

Reviewer 2: ```suggestion
See the following topic for instructions if you are using a third-party provider that only exposes `backend.run`:
```

-  [Update code that calculates expectation values](qiskit-runtime-examples#estimator-algorithm)
-  [Common use cases (basic, parameterized, and dynamic circuits)](qiskit-runtime-use-case)

> ### Step 3a: Replace any `backend.run` options with `qiskit_ibm_runtime` options.

Reviewer 2: ```suggestion
### Step 3a: Replace any `backend.run` options with `qiskit_ibm_runtime` options
```

See the following topics for instructions:

- [Migrate options](qiskit-runtime-options)
- Common use cases [Options section](qiskit-runtime-use-case#migrate-options)

> ### Step 3b: If you want to access a 3rd-party provider that only exposes `backend.run`, you can still follow a primitive-compatible your workflow using the Qiskit backend primitives

Reviewer 2: I think maybe a word is missing (or is there an extra word?) in this header?

Reviewer: +1 on this. How about

```
Step 3: Replace the call to backend.run with a call to primitives

See these topics for instructions if you are using `qiskit-ibm-runtime`:

- Update code that performs circuit sampling
- Update code that calculates expectation values
- Common use cases (basic, parameterized, and dynamic circuits)
- Replace any backend.run options with Qiskit Runtime primitive options
    - Migrate options
    - Common use cases Options section

See the following topic for instructions if you are using a 3rd-party provider that only exposes backend.run:

...
```

See the following topic for instructions:

- [Migrate using backend primitives](qiskit-backend-primitives)

## Information to share with your users

> There is already a collection of migration guides for users to help with the following tasks:

Reviewer 2: ```suggestion
The following migration guides can help users with the following tasks:
```

> - [Transition to the Qiskit Runtime provider](/migration-guides/qiskit-runtime)

Reviewer: Does this just link to the top of this same page?

Reviewer: ok... but why is it called `Transition to the Qiskit Runtime provider`? Is it meant to say transitioning from `qiskit-ibmq-provider` / `qiskit-ibm-provider` to `qiskit-ibm-runtime` ?

Author: I'm not sure why it's called like this, but given that the title of the page is now "Migrate to primitives", and that this is self-referencing, I think this one would make more sense:

```suggestion
- [Transition from `backend.run` to primitives](/migration-guides/qiskit-runtime)
```

Reviewer 3: I agree that your suggested title makes more sense.  Currently, the title of this file is actually "Overview", which is clearly a mistake, so I'll update it. You can verify that I named it correctly.

Reviewer 3: Also, @abbycross told me that we're going to use backend.run without (), so I updated those in this file as well.

> - [Update user code to the V2 primitives interface](/migration-guides/v2-primitives)

Reviewer: I'm not sure if this is needed... given it's a migration guide for backend.run to primitives, readers should not have used v1 primitives.
